Media Text Block: Image size option for featured image

akasunil opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Resolution size support for the featured image is missing in the media text block.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
When the featured image is used in media text block, the `ResolutionTool` is not currently available.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
This PR implement `ResolutionTool` support for featured image. Along with that, cleaning up some issues/bugs that was already present trunk and discovered while working on this PR. 
Following are the fixes,
- Unset the `Use Featured Image` on click of the `Reset` option in the toolbar
- Highlight `Use Featured Image` option on selection.
- New media should load the chosen resolution size if the resolution option is set and the media changes.

## Testing Instructions
 1. Add Media-Text block
 2. Select `Use Featured Image` option from toolbar of Media-Text  block
 3. Look for `Resolution` dropdown in Media-Text block controls
 4. Select different size in Resolution dropdown, Notice the change of image size in editor and on front-end
 5. Update featured image from post setting sidebar, and make sure Media-Text block image get updated in editor and on front-end with same resolution that was previously selected.
 
 **Check the backward compatibility,**
 - Move to trunk, Add some media blocks with different size of image, Save it. 
 - Move to this PR, and make sure that blocks loads without any invalidated markup.

## Screenshots or screencast <!-- if applicable -->
 

[Edit-Post-“Hello-world-”-‹-gutenberg-—-WordPress (4).webm](https://github.com/user-attachments/assets/aead984b-8658-4549-8993-2e7e35073197)

akasunil committed:
Add resolution size support for featured image in media text block

akasunil committed:
Set selected size media url on media change

akasunil committed:
Unset useFeaturedImage on reset click

akasunil committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into enhancement/resolution-option-for-featured-image-media-text-block

akasunil committed:
Fix liniting errors

aaronrobertshaw commented on packages/block-library/src/media-text/edit.js:
```diff
@@ -103,9 +104,8 @@ function attributesFromMedia( {
 		if ( mediaType === 'image' ) {
 			// Try the "large" size URL, falling back to the "full" size URL below.
 			src =
-				media.sizes?.large?.url ||
-				// eslint-disable-next-line camelcase
-				media.media_details?.sizes?.large?.source_url;
+				media.sizes?.[ mediaSizeSlug ]?.url ||
+				media?.media_details?.sizes?.[ mediaSizeSlug ]?.source_url;
```
This code no longer matches the inline comment above. 

I assume the intent of this change is so that any specified `mediaSizeSlug` is used however now it won't prefer large instead of full as the unchanged comment suggests.

aaronrobertshaw reviewed (commented):
Thanks for putting this together @akasunil 👍 

It looks like this PR makes an unintended change to the way image urls fallback i.e. they no longer prefer the `large` size rather than `full` if `mediaSizeSlug` isn't set. 

Can you also update the PR test instructions to cover this aspect of the PR's changes?

akasunil committed:
Update comment text

akasunil committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into enhancement/resolution-option-for-featured-image-media-text-block

aaronrobertshaw reviewed (commented):
I should have been clearer with my [last review comment](https://github.com/WordPress/gutenberg/pull/67464#discussion_r1866947441), sorry 🙏 

The unexpected change of behaviour was the real problem not the comment. The original behaviour should persist unless there was a good reason to change it. 

In making changes to existing code or styles it's really helpful to keep [Chesterton's Fence](https://fs.blog/chestertons-fence/) in mind. The TL;DR on that principle is we need to understand why something is in place before we change or move it to avoid unintended side effects.

akasunil commented:
> The unexpected change of behaviour was the real problem not the comment. The original behaviour should persist unless there was a good reason to change it.

Please checkout the attached video to see the problem that exists in the trunk.

[Edit-Post-‹-gutenberg-—-WordPress.webm](https://github.com/user-attachments/assets/516d9ddc-9e09-4bf1-809b-42e237f5e0c2)


If resolution size is set to `thumbnail`, and we change the media image, New media image should load selected resolution size instead of  `large` size.

aaronrobertshaw commented:
> Please checkout the attached video to see the problem that exists in the trunk.

It might help if you can articulate the steps to reproduce an issue and what reviewers should be looking for.

The same should apply to my last review too even though I was short on time.

> If resolution size is set to thumbnail, and we change the media image, New media image should load selected resolution size instead of large size.

The unexpected change in behaviour wasn't the desire to use any selected resolution but the change in behaviour when a resolution hasn't been selected. With no resolution selected, instead of preferring the large size, falling back to full, the code in question now returns `undefined`.

> Default image size is changed from large to full if mediaSizeSlug isn't available, as default resolution option is full.

This test instruction clearly states a change in behaviour for default image size because something new that is being added had a different default. That's a good sign something isn't right.

akasunil committed:
Use selected media size from setting

akasunil committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into enhancement/resolution-option-for-featured-image-media-text-block

akasunil commented:
Good day, @aaronrobertshaw I apologize for my lack of explanation.

> The code in question now returns `undefined`.

[This](https://github.com/WordPress/gutenberg/pull/67464/files#diff-799bc423023028d1e733982e9d106233470b262e34737b34c559ab7b14e001d7L127) takes care of that. If it is left undefined, the original image URL will be used. The same goes for large sizes as previously implemented.

I prefer `full` size as the default size, Because it is what all other blocks ( such as the [image](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/image/edit.js#L240-L247 ), [cover](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/cover/edit/index.js#L209-L231), [featured-image](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/post-featured-image/edit.js#L44-L48 ) ) do. Keeping the consistency with other media blocks. 

> Use "large" size when adding local images. This is a must-have fix, otherwise we force everybody to download huge images.

I suppose this is why `large` size was set as the default; find out from [here](https://github.com/WordPress/gutenberg/pull/11377#issuecomment-437348279) and PR https://github.com/WordPress/gutenberg/pull/11682 that implements it

aaronrobertshaw commented:
> I prefer full size as the default size, Because it is what all other blocks ( such as the [image](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/image/edit.js#L240-L247%E2%80%A8), [cover](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/cover/edit/index.js#L209-L231), [featured-image](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/post-featured-image/edit.js#L44-L48%E2%80%A8) ) do. Keeping the consistency with other media blocks.

I appreciate the consistency across blocks angle but there is a commitment to backwards compatibility. If we're changing the end result of the save function, for instance due to changing the image src url, that might require a deprecation. 

On that note, have you tested blocks created before this PR on trunk and ensured they are still valid after the update to `save`? The test instructions already need updating after the latest changes, maybe they could include the steps to take to ensure backwards compatibility?

> I suppose this is why large size was set as the default; find out from https://github.com/WordPress/gutenberg/pull/11377#issuecomment-437348279 and PR https://github.com/WordPress/gutenberg/pull/11682 that implements it

There does appear to have been a valid reason for the existing behaviour. This is why [Chesterton's Fence](https://fs.blog/chestertons-fence/) is such a great concept to keep in mind.

akasunil commented:
> There does appear to have been a valid reason for the existing behaviour.

But this is something that possible from theme.json now, we have setting for `imageDefaultSize`. 

> On that note, have you tested blocks created before this PR on trunk and ensured they are still valid after the update to `save`? The test instructions already need updating after the latest changes, maybe they could include the steps to take to ensure backwards compatibility?

Yes, i have tested it, Here what i did, 
- In trunk, Added media text block with image size no-selection, full-size, thumbnail-size and saved it.
- Switched to PR, all block loaded with valid content, No change in image sizes.

ill update the PR description.

akasunil committed:
Fall back to default when selected size media not available

akasunil committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into enhancement/resolution-option-for-featured-image-media-text-block

akasunil commented:
> I take it you mean possible via [block editor settings](https://github.com/WordPress/gutenberg/blob/32cbb04b765d8d8df959642c0d01326327e5ff41/lib/experimental/class-wp-rest-block-editor-settings-controller.php#L255)?

yes, that one.

aaronrobertshaw reviewed (commented):
Thanks for the persistence here @akasunil 👍 

There still seems to be a gap in the fallback logic around the featured image size.

If I select a featured image that does not have the default `large` size it falls back to the `thumbnail` option in the UI but the image doesn't match that. 


https://github.com/user-attachments/assets/16225850-f066-42e2-9f99-b4accc1082b0



Similarly, if you have a featured image selected and set the resolution to `large` or any size really, then switch the featured image to one that doesn't have that size, the resolution control again displays the `thumbnail` size as selected.


https://github.com/user-attachments/assets/eac45515-54e1-4464-9c84-b7de29036249


FYI, I'll be AFK for most of the next few weeks. It might be a while before I can revisit this sorry. Feel free to ping other reviewers for fresh eyes and to keep the momentum going!

akasunil committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into enhancement/resolution-option-for-featured-image-media-text-block

akasunil committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into enhancement/resolution-option-for-featured-image-media-text-block

akasunil committed:
Fix featured image default size issue

akasunil commented:
Thank you @aaronrobertshaw for your feedback.

> There still seems to be a gap in the fallback logic around the featured image size.

I fixed this for normal images, missed to check for featured image. Anyway, this is fixed now for featured images too.

aaronrobertshaw reviewed (commented):
@akasunil the issues flagged in my last review, to do with having the correct size preselected for feature images etc, don't appear to have been resolved.

Some simple steps to replicate are:

1. Add media and text block
2. Add featured image that has `large` size
3. Select media and text block and choose to use featured image
4. Select the `large` size for the image in that block
5. Change the featured image for the post to one that doesn't have the `large` size
6. Reselect the media and text block, switch to the block inspector and note the size incorrectly says `thumbnail`.

Additionally, there are still other quirks like the featured image in the block defaulting to a different size than normal images. Shouldn't those be consistent?

1. Add two media and text blocks
2. Select a featured image for the post that has a `large` and `full` sizes
3. Select the first media and text block and choose for it to display the featured image
4. Note that this block's image defaults to `Full size`
5. Select the second media and text block
6. Choose to use an image from the media library and select the same image as used for the featured image
7. Note that this image uses the `Large` size by default.

Did you test these after my last review? If so, are you seeing different behaviour?


https://github.com/user-attachments/assets/4fbd89b6-3d00-497e-8276-1f57d1e13fec



> It might be a while before I can revisit this sorry. Feel free to ping other reviewers for fresh eyes and to keep the momentum going!

I am out of bandwidth to keep reviewing this PR, sorry 🙏

To help other reviewers, with fresh eyes,   it would help to keep the PR test instructions up-to-date. In particular, add some instructions around testing switching between images with different available sizes.

akasunil committed:
Resolved conflict and sync with trunk

akasunil committed:
Merge branch 'trunk' of personal.github.com:WordPress/gutenberg into enhancement/resolution-option-for-featured-image-media-text-block

akasunil committed:
Merge branch 'trunk' of personal.github.com:WordPress/gutenberg into enhancement/resolution-option-for-featured-image-media-text-block

akasunil committed:
Merge branch 'trunk' of personal.github.com:WordPress/gutenberg into enhancement/resolution-option-for-featured-image-media-text-block

akasunil committed:
Fix mediaSizeSlug variable error

akasunil committed:
Fix duplicate image variable error

akasunil committed:
Fix featured image default size issue

akasunil commented:
> How do I check if an image in the media library doesn't have a large size? Aren't these sizes generated? Do I need to delete some files?

I'm not sure, but i guess it depend on original size image image. try using large image like 1200 or 1600px.

akasunil commented:
Thank you @carolinan, I have fixed the build issue and also feedback given by @aaronrobertshaw, You can review this PR now.
